pkg/controllers/management/certsexpiration/certsexpiration.go

```diff
@@ -0,0 +1,176 @@
+package certsexpiration
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/rancher/kontainer-engine/cluster"
+	"github.com/rancher/rancher/pkg/clustermanager"
+	"github.com/rancher/rancher/pkg/controllers/management/clusterprovisioner"
+	rkecluster "github.com/rancher/rke/cluster"
+	"github.com/rancher/rke/pki"
+	v1 "github.com/rancher/types/apis/core/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/cert"
+)
+
+type Controller struct {
+	ClusterLister   v3.ClusterLister
+	ClusterClient   v3.ClusterInterface
+	ClusterManager  *clustermanager.Manager
+	NamespaceClient v1.NamespaceInterface
+	CoreV1          v1.Interface
+}
+
+func Register(ctx context.Context, management *config.ManagementContext, manager *clustermanager.Manager) {
+	c := &Controller{
+		ClusterLister:   management.Management.Clusters("").Controller().Lister(),
+		ClusterClient:   management.Management.Clusters(""),
+		NamespaceClient: management.Core.Namespaces(""),
+		CoreV1:          management.Core,
+		ClusterManager:  manager,
+	}
+	management.Management.Clusters("").AddHandler(ctx, "certificate-expiration", c.sync)
+}
+
+func (c Controller) sync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if key == "" || cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if cluster.Spec.RancherKubernetesEngineConfig == nil {
+		return nil, nil
+	}
+	certsExpInfo := map[string]v3.CertExpiration{}
+
+	cluster, err := c.ClusterLister.Get("", key)
+	if err != nil {
+		return nil, err
+	}
+	certBundle, err := c.getClusterCertificateBundle(cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+	for certName, certObj := range certBundle {
+		info, err := getCertExpiration(certObj.CertificatePEM)
+		if err != nil {
+			continue
+		}
+		certsExpInfo[certName] = info
+	}
+	cluster.Status.CertificatesExpiration = certsExpInfo
+	return c.ClusterClient.Update(cluster)
+
+}
+
+func (c Controller) getClusterCertificateBundle(clusterName string) (map[string]pki.CertificatePKI, error) {
+	// cluster has a state file ?
+	store := clusterprovisioner.NewPersistentStore(c.NamespaceClient, c.CoreV1)
```

Comment:
why are we getting a new store every time? would it help to just have it in controller definition?